fix: improve our docker config

aronwk-aaron opened the pull request:
greatly simplify docker and automaically build container and push to ghcr.io
fixes #570 
fixes #418
fixes #417

- Removes account manager as a submodule and repalces it with the nexus dashboard docker container
- Removed all the customs dockerfiles for things like mariadb
- Removed the need for users to build the image locally
- Change port config name in each servers to specific to what sever it is as to avoid env var name collisions
- Tested that the docker image works by locally building it and connecting to it with a client

Xiphoseer commented on Dockerfile:
```diff
@@ -0,0 +1,35 @@
+
+COPY --from=build /app/build/*Server /app/
+COPY --from=build /app/build/*.ini /app/default-configs/
+COPY --from=build /app/build/*.so /usr/lib/
```
While this is technically correct, without another fix, this will be missing the `libmariadb.so`.

Xiphoseer commented on README.md:
````diff
@@ -347,6 +348,43 @@ certutil -hashfile <file> SHA1
+```
+You will need to replace the `/path/to/`'s to reflect the paths on your host.
+
+Any config option in the `.ini`'s can be overridden with environmental variables: Ex: `log_to_console=1` from `shared_config.ini` would be overidden like  `-e LOG_TO_CONSOLE=0`
````
Maybe mention that all servers will pick it up, so service specific settings may have trouble and should use `{world,chat,auth,master}config.ini`

Xiphoseer replied:
Ah neat, you fixed the cases where that would have been an issue (port)

aronwk-aaron replied:
oh yeah, I think i put that in the PR description when opening it for review

EmosewaMC commented on Dockerfile:
```diff
@@ -0,0 +1,35 @@
+
+COPY . /app/
+
+RUN sed -i 's/MARIADB_CONNECTOR_COMPILE_JOBS__=1/MARIADB_CONNECTOR_COMPILE_JOBS__=2/' /app/CMakeVariables.txt 
```
This should be looking for `MARIADB_CONNECTOR_COMPILE_JOBS__=` instead of `MARIADB_CONNECTOR_COMPILE_JOBS__=1` as the latter will not be found if we on the master repo change it from the default

aronwk-aaron replied:
will need to modify the sed to just replace the whole line then

EmosewaMC commented on Dockerfile:
```diff
@@ -0,0 +1,35 @@
+
+RUN sed -i 's/MARIADB_CONNECTOR_COMPILE_JOBS__=1/MARIADB_CONNECTOR_COMPILE_JOBS__=2/' /app/CMakeVariables.txt 
+
+RUN mkdir build
+RUN cd build && \
+	cmake .. && \
+	make -j2
```
Can this not be replaced with `./build.sh -j2`?  That file does the same work as is done here

Xiphoseer commented on entrypoint.sh:
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
+
+cp -n /app/default-configs/* /app/configs/
```
Why copy in the entrypoint? Can you remove this and put master at the ENTRYPOINT directly?

aronwk-aaron replied:
this puts the files on disk. Which will be mounted, and make these files persistent.
This gives users the option to edit the configs on disk instead of using env vars if they so wish.

Same with vanity, allows users to edit them on the mounted disk.

I'm not sure if there's a better way to do this.
But I always like when images do this and I don't need to go manually grab default configs from whereever to put on my host disk

Xiphoseer replied:
Some thoughts:
- I would argue this cp -n is worse than none by default because it doesn't diff, so once you upte you have stale config sitting there preventing a copy and you need to merge by hand again. I do see the appeal of default config files showing up on their own though.
- I'm worried that this is another place added to the already complex master startup that does magic. For k8s for example I need to rely on being able to replace entrypoint with WorldServers so this script must not have load bearing (fatal when skipped) side effects

aronwk-aaron replied:
Stale configs is just part of it unfortunately. There is no remedy other than making the server possibly update the configs with new setting during startup and providing a sane default (which we could do that since we'll have the latest default configs in the container always)

All the necessary configs already exist and any new configs should only be superfluous with the default behavior that the setting would change being what happened before it was added.

This is 100% just a way to get a bootstrap on a fresh container.
Users of docker know they can modify the entrypoint in their run command if they need to do so, which I've personally rarely had to do. I'm assuming you're already modifying the entrypoint in k8's since you're deploying individual servers to separate pods (I think that terminology is correct)

aronwk-aaron replied:
> There is no remedy other than making the server possibly update the configs with new setting during startup and providing a sane default

Having the configs be writable may be a concern as well, but that's a different discussion imo

Also, could just forsake this all and force the user to either do everything via env vars or know they need to bind mount the configs and the go grab them from the repo

And we should probably remove the dependency for the configs files needing to exist at all as well, but that can be a separate PR.

This is all just to get a MVP for docker without any server changes

Xiphoseer reviewed (dismissed):
with nits

EmosewaMC commented on README.md:
````diff
@@ -347,6 +348,43 @@ certutil -hashfile <file> SHA1
+This assumes that you have a database deployed to your host or in another docker container.
+
+A basic deployment of this contianer would look like:
+```sh
+# example docker contianer deployment
+docker run -it \
+    -v /path/to/configs/:/app/configs \
+    -v /path/to/logs/:/app/logs \
+    -v /path/to/dumps/:/app/dumps \
+    -v /path/to/res:/app/res:ro \
+    -v /path/to/resServer:/app/resServer \
+    -e DLU_CONFIG_DIR=/app/configs \
+    -e DUMP_FOLDER=/app/dumps \
+    -p 1001:1001/udp \
+    -p 2005:2005/udp \
+    -p 3000-3300:3000-3300/udp \
+ghcr.io/darkflameuniverse/darkfameserver:latest
+```
````
imo this should be moved to a shell.example file for people to edit instead in one place to just cut the middle man step of needing to copy this text into a file and instead of copy pasting a file and renaming it.  Plus then we can reference the example file by name and not have a huge docker container command in the readme

aronwk-aaron replied:
this is standard to have in the readme for containers

EmosewaMC replied:
Can't say I agree but alright

Xiphoseer commented on Dockerfile:
```diff
@@ -0,0 +1,35 @@
+FROM gcc:latest as build
```
Just noticed: this should be a number, not `latest`, otherwise it'll break on its own down the line. Can be just the major semver component though, to get a bit of updates automatically.

aronwk-aaron replied:
yep, good catch. I meant to do this

EmosewaMC commented on docker-compose.yml:
```diff
@@ -1,104 +1,147 @@
+      # master config options
+      - MASTER_IP=${MASTER_IP:-localhost}
+      - MASTER_PORT=${MASTER_PORT:-2000}
+      - WORLD_PORT_START=${WORLD_PORT_START:-3000}
+      - USE_SUDO_AUTH=${USE_SUDO_AUTH:-0}
+      - USE_SUDO_CHAT=${USE_SUDO_CHAT:-0}
+      - USE_SUDO_WORLD=${USE_SUDO_WORLD:-0}
+      - PRESTART_SERVERS=${PRESTART_SERVERS:-1}
+      # world config options
+      - SOURCE=${SOURCE:-https://github.com/DarkflameUniverse/DarkflameServer}
+      - CHAT_SERVER_PORT=${CHAT_SERVER_PORT:-2005}
+      - DISABLE_CHAT=${DISABLE_CHAT:-0}
+      - PHYS_SPATIAL_PARTITIONING=${PHYS_SPATIAL_PARTITIONING:-1}
+      - PHYS_SP_TILESIZE=${PHYS_SP_TILESIZE:-102}
+      - PHYS_SP_TILECOUNT=${PHYS_SP_TILECOUNT:-24}
+      - DISABLE_EXTRA_BACKPACK=${DISABLE_EXTRA_BACKPACK:-0}
+      - DISABLE_VANITY=${DISABLE_VANITY:-0}
+      - SOLO_RACING=${SOLO_RACING:-0}
+      - DISABLE_ANTI_SPEEDHACK=${DISABLE_ANTI_SPEEDHACK:-0}
+      - CHECK_FDB=${CHECK_FDB:-0}
+      - CLASSIC_SURVIVAL_SCORING=${CLASSIC_SURVIVAL_SCORING:-0}
+      - PETS_TAKE_IMAGINATION=${PETS_TAKE_IMAGINATION:-1}
+      - DISABLE_DROPS=${DISABLE_DROPS:-0}
+      - HARDCORE_MODE=${HARDCORE_MODE:-0}
+      - HARDCORE_DROPINVENTORY_ON_DEATH=${HARDCORE_DROPINVENTORY_ON_DEATH:-1}
+      - HARDCORE_USCORE_ENEMIES_MULTIPLIER=${HARDCORE_USCORE_ENEMIES_MULTIPLIER:-2}
+      - HARDCORE_LOSE_USCORE_ON_DEATH_PERCENT=${HARDCORE_LOSE_USCORE_ON_DEATH_PERCENT:-10}
+      - ALLOW_NAMEPLATE_OFF=${ALLOW_NAMEPLATE_OFF:-0}
+      - LOG_IP_ADDRESSES_FOR_ANTI_CHEAT=${LOG_IP_ADDRESSES_FOR_ANTI_CHEAT:-1}
+      - HELP_0_SUMMARY=${HELP_0_SUMMARY:-Got an issue?}
+      - HELP_1_SUMMARY=${HELP_1_SUMMARY:-Stuck loading?}
+      - HELP_2_SUMMARY=${HELP_2_SUMMARY:-Missing features?}
+      - HELP_3_SUMMARY=${HELP_3_SUMMARY:-Get smashed?}
+      - HELP_4_SUMMARY=${HELP_4_SUMMARY:-Want to contribute?}
+      - HELP_0_DESCRIPTION=${HELP_0_DESCRIPTION:-Go to the DarkflameServer repository on GitHub to view issues and discussions about the server emulator!<br/><br/><a href="https://github.com/DarkflameUniverse/DarkflameServer"><font color="#0000EE">Click Here to go there!</font></a><br/><br/>}
+      - HELP_1_DESCRIPTION=${HELP_1_DESCRIPTION:-Try switching networks, using a VPN, or using your phone's hotspot to resolve the issue.<br/><br/>}
+      - HELP_2_DESCRIPTION=${HELP_2_DESCRIPTION:-While DarkflameServer is a mostly complete emulator, there are still some features that aren't implemented. You can track these on the GitHub issues page.<br/><br/>}
+      - HELP_3_DESCRIPTION=${HELP_3_DESCRIPTION:-Skill issue!<br/><br/>}
+      - HELP_4_DESCRIPTION=${HELP_4_DESCRIPTION:-Visit Discussions on the DarkflameServer GitHub page<br/>to ask questions and collaborate with other devs!<br/><br/>}
+      - ALLOW_PLAYERS_TO_SKIP_CINEMATICS=${ALLOW_PLAYERS_TO_SKIP_CINEMATICS:-0}
```
I'm still not a fan of this huge copy paste here which is prone to errors, forgetting to update and is just messy overall.  Is there really no other possible way to do this more efficiently?

aronwk-aaron replied:
Just going to remove most of them and leave only the necessary ones

Xiphoseer commented on Dockerfile:
```diff
@@ -0,0 +1,35 @@
+	cmake .. && \
+	make -j2
+
+FROM gcc:latest as runtime
```
same here, consider `debian:12`

aronwk-aaron replied:
Will try this

EmosewaMC reviewed (dismissed):
Because we are changing the port option name, we need to add a block of code to the CMakeLists.txt file to update a line that starts with `port=` and updates it to say auth_port etc.  This is a breaking change otherwise as some users may have had to change some of the default ports for things like WorldServer start or where MasterServer resides due to port collisions

aronwk-aaron committed:
greatly simplify docker

aronwk-aaron committed:
back to gcc since I can't the libs to load properly

EmosewaMC commented on Dockerfile:
```diff
@@ -0,0 +1,35 @@
+# needed as the container runs with the root user
+# and therefore sudo doesn't exist
+ENV USE_SUDO_AUTH=0
+RUN chmod +x entrypoint.sh
```
Can entrypoint.sh not just have execute perms by default?

aronwk-aaron replied:
I tried setting it to +x on my host and then building the container. And they did not work. So yes, this is required

aronwk-aaron replied:
https://stackoverflow.com/questions/27281965/docker-file-chmod-on-entrypoint-script

EmosewaMC commented on README.md:
```diff
@@ -347,6 +348,43 @@ certutil -hashfile <file> SHA1
+## Standalone
+
+For standalone deployment, you can use the image provided via Github's Container Repository:
+`ghcr.io/darkflameuniverse/darkfameserver`
```
`darkfameserver`

EmosewaMC commented on entrypoint.sh:
```diff
@@ -0,0 +1,6 @@
+cp -n /app/default-configs/* /app/configs/
+mkdir -p /app/vanity/
+cp -n /app/default-vanity/* /app/vanity/
+cd /app/ && ./MasterServer
```
Should this file have `set -e` at the start?

aronwk-aaron replied:
Potentially, won't hurt to add it

aronwk-aaron committed:
fix typo
add set -e to entrypoint
better copy of entrypoint.sh and use proper entrypoint

aronwk-aaron committed:
Merge branch 'main' into fix--remove-the-wacrimes-from-our-docker-config

aronwk-aaron committed:
use debian instead of gcc for runtime
comment and organize it a bit
drop gcc to 12 since we are using debian 12 as well

aronwk-aaron committed:
explicitly include mariadb libs

aronwk-aaron commented:
re-tested that this works and connected to a server running in docker with a client on the same host

aronwk-aaron committed:
Make the server not crash in the case we are using only env-vars
make the dockerfile have configs in the expected location incase of bypassing entrypoint.sh

aronwk-aaron commented:
`s/crash/exit/` on the last commit message

aronwk-aaron committed:
remove unneede var from example, since it's in the container now

aronwk-aaron committed:
coments to dockerfile

aronwk-aaron committed:
Revert master server changes

EmosewaMC commented on dMasterServer/MasterServer.cpp:
```diff
@@ -81,43 +81,18 @@ int main(int argc, char** argv) {
+	if (!dConfig::Exists("chatconfig.ini")) LOG("Couldnt find chatconfig.ini");
+	if (!dConfig::Exists("masterconfig.ini")) LOG("Couldnt find masterconfig.ini");
+	if (!dConfig::Exists("sharedconfig.ini")) LOG("Couldnt find sharedconfig.ini");
+	if (!dConfig::Exists("worldconfig.ini")) LOG("Couldnt find worldconfig.ini");
```
Instead of removing checks like this please instead consider checking if there is an env var set which allows the bypass of the check inatead

aronwk-aaron replied:
yep, already reveted these chagnes and am going to consider these changes in a different PR

EmosewaMC commented on dMasterServer/MasterServer.cpp:
```diff
@@ -81,43 +81,18 @@ int main(int argc, char** argv) {
-		return EXIT_FAILURE;
-	}
+	uint32_t clientNetVersion = 171022;
+	GeneralUtils::TryParse(Game::config->GetValue("client_net_version"), clientNetVersion);
```
Same here.

aronwk-aaron committed:
Merge branch 'main' into fix--remove-the-wacrimes-from-our-docker-config

aronwk-aaron committed:
Merge branch 'main' into fix--remove-the-wacrimes-from-our-docker-config

aronwk-aaron committed:
Resolve conflicting port options between chat, master, and world
move chat_server_port to shared since it's used by world and chat

aronwk-aaron committed:
Don't error if file does not exists when updating a config option
move update before and use bin dir var
